jaxlib/mosaic/gpu/mosaic_gpu_comm.h

```diff
@@ -0,0 +1,82 @@
+/* Copyright 2025 The JAX Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef JAXLIB_MOSAIC_GPU_COMM_H_
+#define JAXLIB_MOSAIC_GPU_COMM_H_
+
+#include <dlfcn.h>
+#include <string.h>
+
+#include "third_party/gpus/cuda/include/cuda.h"
+#include "third_party/gpus/cuda/include/cuda_runtime_api.h"
+
+#define NVSHMEM_SUCCESS 0
+#define NVSHMEM_LIB_SONAME "libnvshmem_host.so.3"
+
+namespace mosaic {
+namespace gpu {
+
+#define NVSHMEM_SET_FN(FnName)                                            \
+  FnName = reinterpret_cast<decltype(FnName)>(dlsym(library, #FnName));   \
+  if (!FnName) {                                                          \
+    fprintf(stderr, #FnName " not available in this library.");           \
+    abort();                                                              \
+  }
+
+class NvshmemApi {
+ public:
+  // Returns a default NvshmemApi for a current process.
+  // NvshmemApi follows the Singleton design pattern
+  static NvshmemApi& Default() {
+    static NvshmemApi instance;
+    return instance;
+  }
+
+  int cumodule_int(CUmodule module) {
+    return nvshmemx_cumodule_init(module);
+  }
+
+  void barrier_all_on_stream(cudaStream_t stream) {
+    nvshmemx_barrier_all_on_stream(stream);
+  }
+
+  NvshmemApi(NvshmemApi const&)     = delete;
+  void operator=(NvshmemApi const&) = delete;
+
+ private:
+  NvshmemApi() {
+    const char* env_value = getenv("NVSHMEM_LIBRARY_PATH");
+    const char* libnvshmem_path =
+      env_value && strlen(env_value) > 0 ? env_value : NVSHMEM_LIB_SONAME;
+    void* library = dlopen(libnvshmem_path, RTLD_LAZY);
+    if (library == nullptr) {
+      fprintf(stderr, "Failed to open %s library: %s", libnvshmem_path, dlerror());
+      abort();
+    }
+
+    // Initialize supported NVSHMEM host API
+    NVSHMEM_SET_FN(nvshmemx_cumodule_init)
+    NVSHMEM_SET_FN(nvshmemx_barrier_all_on_stream)
+  }
+
+  // Dlopened NVSHMEM API
+  int (*nvshmemx_cumodule_init)(CUmodule);
+  int (*nvshmemx_barrier_all_on_stream)(cudaStream_t);
+};
+
+}  // namespace gpu
+}  // namespace mosaic
+
+#endif  // JAXLIB_MOSAIC_GPU_COMM_H_
```

Comment:
Could we just include this header in the runtime file? It's not like we're planning to include it from anywhere else as far as I can tell? I'd also strip it down to not use C++ features, just to make sure the .so we generate for the runtime won't be annoying to package

Reply:
I am pretty sure we use a C++ compiler, but C++ has a messy ABI, and we need to interface with those functions from inside the kernel so ABI matters

Reply:
Using:
```
extern "C" {
   void foo();
}
```
for solving that issue not to get the C++ mangling for functions. Probably that should be enough here.

Reply:
It's not just about the symbol names the `.so` should export, but it's also about the dependency on `libc++`/`libcxx` which can be a pain (and this depends on whether you include C++ headers or not). Anyway, let's try and see.